pkg/remotecluster/remotecluster.go

```diff
@@ -0,0 +1,202 @@
+
+	ClusterID             string
+	ClientSet             kubernetes.Interface
+	podInformer           cache.SharedIndexInformer
+	networkPolicyInformer cache.SharedIndexInformer
+
+	eventChanMutex *sync.Mutex
+	eventChan      chan *Event
+}
+
+type EventType string
+
+const (
+	AddEvent    EventType = "Added"
+	DeleteEvent EventType = "Deleted"
+	UpdateEvent EventType = "Updated"
+)
+
+type ObjectType string
+
+const (
+	NetworkPolicy ObjectType = "np"
+	Pod           ObjectType = "pod"
+)
+
+type Event struct {
+	Cluster *RemoteCluster
+	Type    EventType
+	ObjType ObjectType
+	Objs    []interface{}
+}
+
+func New(clusterID string, clientSet kubernetes.Interface) *RemoteCluster {
+
+	factory := informers.NewSharedInformerFactory(clientSet, time.Hour*24)
+	podInformer := factory.Core().V1().Pods().Informer()
+	networkPolicyInformer := factory.Networking().V1().NetworkPolicies().Informer()
+
+	resourceWatcher := &RemoteCluster{
+		stopCh:                make(chan struct{}),
+		ClusterID:             clusterID,
+		ClientSet:             clientSet,
+		podInformer:           podInformer,
+		networkPolicyInformer: networkPolicyInformer,
+		eventChanMutex:        &sync.Mutex{},
+	}
+
+	podInformer.AddEventHandler(resourceWatcher)
+	networkPolicyInformer.AddEventHandler(resourceWatcher)
+
+	return resourceWatcher
+}
+
+func (rw *RemoteCluster) HasSynced() bool {
+	return rw.podInformer.HasSynced() &&
+		rw.networkPolicyInformer.HasSynced()
+}
+
+// Stop will stop the running informers
+func (rw *RemoteCluster) Stop() {
+	close(rw.stopCh)
+}
+
+func (rw *RemoteCluster) Stopped() bool {
+	select {
+	case _, ok := <-rw.stopCh:
+		// channel is not ok to read, means it's closed now
+		return !ok
+	default:
+		// channel was ok to read, but had no messages
+		return false
+	}
+}
+
+func (rw *RemoteCluster) Run(onSyncDoneFunc func(resourceWatcher *RemoteCluster)) {
+	go rw.podInformer.Run(rw.stopCh)
+	go rw.networkPolicyInformer.Run(rw.stopCh)
+
+	go func() {
+		if !cache.WaitForCacheSync(rw.stopCh, rw.podInformer.HasSynced) {
+			klog.Warning("Timed out waiting for pod informer to sync")
+		}
+
+		if !cache.WaitForCacheSync(rw.stopCh, rw.networkPolicyInformer.HasSynced) {
+			klog.Warning("Timed out waiting for NetworkPolicy informer to sync")
+		}
+
+		if onSyncDoneFunc != nil {
+			onSyncDoneFunc(rw)
+		}
+	}()
+}
+
+func (rw *RemoteCluster) GetPods() []interface{} {
+	return rw.podInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) GetNetworkPolicies() []interface{} {
+	return rw.networkPolicyInformer.GetStore().List()
+}
+
+func (rw *RemoteCluster) SetEventChannel(eventChan chan *Event) {
+	rw.eventChanMutex.Lock()
+	rw.eventChan = eventChan
+	rw.eventChanMutex.Unlock()
+}
+
+func (rw *RemoteCluster) enqueueEvent(event *Event) {
+
+	// lock used as memory barrier to make sure eventChannel ref
+	// is synchronized between threads
+	rw.eventChanMutex.Lock()
+	if rw.eventChan != nil {
+		rw.eventChan <- event
+	}
+	rw.eventChanMutex.Unlock()
```

Comment:
defer is great when you have multiple exit paths for the function, but less visual of what the lock is protecting. but just one opinion,

Reply:
IIUC defer works like a finally and guarantees that the unlock will be called no matter what happens before it, hence it seemed safer to me to use that (and i believe it's part of the reason why its so widely used).

Reply:
`defer` is convenient and can avoid latent bugs should (more) exits paths be added in the future but it can also be a double-edge sword as it artificially expands the critical section surface to the end of the function, which may or may not be negligible, and can become worse and worse as the code evolves over time (similar to a synchronized method in Java).  Also @mangelajo makes a good point wrt visualization about what's being protected. In this case I don't think it matters either way. In general, in cases where one is using a critical section to protect access to a data structure, the block should be relatively small and concise without exit paths and thus shouldn't need the safety of a `defer`. If one needs a critical section for atomicity of a sequence of operations, then `defer` may be more appropriate. My 2 cents...